lib/rex/ui/text/shell/history_manager.rb

```diff
@@ -0,0 +1,74 @@
+# -*- coding: binary -*-
+
+module Rex
+module Ui
+module Text
+module Shell
+
+class HistoryManager
+ 
+
+  @@contexts = [{:history_file => Msf::Config.history_file, :name => :msfconsole}]
+  
+  def self.inspect
+    "#<HistoryManager stack size: #{@@contexts.length}>"
+  end
+
+  def self.context_stack
+    @@contexts
+  end
+
+  def self.push_context(history_file: nil, name: nil)
+    return if @@contexts[-1][:name] == name
+    self.clear_readline
+    @@contexts.push({:history_file => history_file, :name => name})
+    if history_file
+      self.set_history_file(history_file)
+    end
+  end
+
+  def self.pop_context
+    if @@contexts.empty?
```

Comment:
If a `#pop_context` operation takes place on an empty stack this would seem to imply that the stack is inconsistent. This should at least be logged as an error.

Reply:
I've added `elog` in case the stack is empty and still `pop_context` is called.